### PR TITLE
Block Sptrsv fixes

### PR DESCRIPTION
**Pull request by jgfouca**
Go back to the vector reduce implementation and fix block divide to have the right order of operations. With these changes, RBILUK+KK produces the same results as RBILUK (without KK).

These changes unfortunately cause a problem I saw earlier to reemerge. The problem is unique to Clang+CUDA+Complex+Release, so I strongly suspect it's a compiler bug. If you use GCC, it works. If you turn off -03 (-O0 and -O1 work), it works. To get around this, I disabled the test for this specific case.

### CODE REVIEW

**Review by cwpearson — CHANGES_REQUESTED** [review]
I suggest:

1. detect when we're in the scenario that's buggy ( compiler + optimization + algorithm choice, if I understand correctly) and throw if someone tries to call.
2. Change the test so it's explicitly GTEST_SKIP in that case with a comment.

**Comment by jgfouca** [comment]
Thanks, @cwpearson . I added some code to the handle to address your first point. I'm not sure I can do the second point because GTEST_SKIP is for when you are skipping an entire TEST; in this case, we are just skipping a small piece of the overall test.

File: sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp [hunk 1]
```diff
@@ -260,14 +327,14 @@ struct SptrsvWrap {
       // A = LU
       // A^-1 = U^-1 * L^-1
```
Comment (lucbv) [inline sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp:328]: The comments here are a bit confusing, especially the line below where you apply the matrix on the left side of `b` which is wrong.
I would say the following:
// Ax = LUx = Lz = b, we use the change of variable z = U*x
// z = L^-1 * b, first we solve for z, storing the result back into b
// x = U^-1 * z, second we solve for x, again storing the result back into b

File: sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp [hunk 2]
```diff
@@ -278,21 +345,44 @@ struct SptrsvWrap {
...
-      KK_KERNEL_REQUIRE_MSG(block_size_ <= MAX_VEC_SIZE,
-                            "Max supported block size for range-policy is 16. Use team-policy alg if you need more.");
-
       Block LU(&buff[0], block_size_, block_size_);
       assign(LU, A);
       KokkosBatched::SerialLU<KokkosBatched::Algo::LU::Blocked>::invoke(LU);
 
       // A = LU
       // A^-1 = U^-1 * L^-1
       // b = (b * U^-1) * L^-1, so do U trsv first
```
Comment (lucbv) [inline sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp:354]: Again this is misleading at best.

**Review by lucbv — COMMENTED** [review]
It looks correct but it would be nice to update the comments so they match the code.

**Comment by cwpearson** [comment]
> Thanks, @cwpearson . I added some code to the handle to address your first point. I'm not sure I can do the second point because GTEST_SKIP is for when you are skipping an entire TEST; in this case, we are just skipping a small piece of the overall test.

Then I can live with printing a message to stderr that says what's going on and why before `continue`

**Comment by lucbv** [comment]
@cwpearson I do not believe that the version of gtest snapshotted in our repo (1.7.0) supports gtest_skip?
We could move to a new gtest as kokkos core is using 1.11.0 and Trilinos is using 1.10.0

**Comment by jgfouca** [comment]
@cwpearson , @nellingwood , I see the one CI job failed:

```
/home/runner/_work/kokkos-kernels/kokkos-kernels/kokkos-kernels/test_common/KokkosKernels_TestUtils.hpp:140: Failure
Expected: ((double)AT1::abs(val1 - val2)) <= ((double)AT3::abs(tol)), actual: 119.541 vs 1.19541e-05
/home/runner/_work/kokkos-kernels/kokkos-kernels/kokkos-kernels/test_common/KokkosKernels_TestUtils.hpp:140: Failure
Expected: ((double)AT1::abs(val1 - val2)) <= ((double)AT3::abs(tol)), actual: 119.541 vs 1.19541e-05
[  FAILED  ] serial.asum_complex_double (0 ms)
```

This is not code that I touched in this PR; is this an expected fail on develop? Maybe it's intermittent?

**Comment by cwpearson** [comment]
The failures are unrelated: https://github.com/kokkos/kokkos-kernels/issues/2005

**Review by cwpearson — APPROVED** [review]

**Comment by jgfouca** [comment]
Everything looks good now. @lucbv , would you mind merging this?

**Review by lucbv — APPROVED** [review]
Yep, looks good to me

**Pull request merged**